Align vocabulary with ISO mDL and Spruce.

timothee-haudebourg opened the pull request:
Here are my efforts to align the current VDL vocabulary with ISO mDL and Spruce VC. For now the only changes are to the JSON-LD context. Once approved I'll update the rest of the spec.

# Changes

- Removed properties that are not specified by ISO mDL
  - `gender` (similar but not equal to ISO mDL `sex`, with a different datatype)
  - `name_national_character` (similar but not equal to ISO mDL `family_name_national_character` and `given_name_national_character`)
- Renamed all properties URL into `camelCase`, following established RDF conventions
- Added properties
  - `familyNameNationalCharacter` of type `http://www.w3.org/2001/XMLSchema#string`
  - `givenNameNationalCharacter` of type `http://www.w3.org/2001/XMLSchema#string`
  - `sex` of type `http://www.w3.org/2001/XMLSchema#unsignedInt`
  - `ageOver18`, `ageOver21`, `ageOver25`, `ageOver62`, `ageOver65` of type `http://www.w3.org/2001/XMLSchema#boolean`, with associated terms in the LD context. Outside of the LD context we can generalize and specify an URL of the form `https://w3id.org/vdl#ageOverNN` for all `NN`.
- Changed datatypes
  - `ageBirthYear`, `ageInYears`, `height`, `weight`: use `http://www.w3.org/2001/XMLSchema#unsignedInt` which limits the range to integers between 0 and 2^32-1 as specified in ISO mDL.
  - `portrait`, `signatureUsualMark`: use `http://www.w3.org/2001/XMLSchema#base64Binary` to represent byte strings. This is the closest we can get to CBOR `bstr` datatype, as there is no correct JSON serialization for this datatype.
  - `birthDate`: use `https://www.rfc-editor.org/rfc/rfc3339#full-date` which refers to the `full-date` production of RFC3339 as specified by ISO mDL. NOTE: this datatype URL is non standard (there is no established RDF datatypes for RFC3339).
  - `portraitCaptureDate`: use `https://www.rfc-editor.org/rfc/rfc3339#date-time` which refers to the `date-time` production of RFC3339 as specified by ISO mDL. NOTE: this datatype URL is non standard (there is no established RDF datatypes for RFC3339).
  - `https://w3id.org/vdl#issueDate` and `https://w3id.org/vdl#expiryDate`: Do not use `http://www.w3.org/2001/XMLSchema#dateTime` where an RFC3339 `date-time` or `full-date` literal is required by ISO mDL. Use the custom datatype `https://w3id.org/vdl#date-time-or-full-date` instead, which can be defined in the VDL specification as a datatype union using OWL:
```ttl
<https://w3id.org/vdl#date-time-or-full-date> rdfs:Datatype ;
  owl:unionOf (
    <https://www.rfc-editor.org/rfc/rfc3339#date-time>
    <https://www.rfc-editor.org/rfc/rfc3339#full-date>
  ) .
```

In the JSON-LD context I also:
  - removed the `id` and `type` term definitions under the `Iso18013DriversLicense` type scoped context, since `id` and `type` schould not occur inside the license (`type` can occur but it is recognized by the toplevel `type` term definition).
  - renamed the `license` term into `driversLicense` to avoid collision with other license kind.

timothee-haudebourg commented:
I've added namespaces in the JSON-LD context definition. The license now should have the following form:
```json
{
  "@type": "Iso18013DriversLicense",
  "namespaces": {
    "org.iso.18013.5.1": {
      "term": "value"
    }
  }
}
```
This allows the definition of terms from other namespaces without risking ambiguities.

In a separate context, I've defined the AAMVA terms. Here is a VC example using both contexts:
```json
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://w3id.org/vdl/v1",
    "https://w3id.org/vdl/aamva/v1"
  ],
  "issuer": "...",
  "issuanceDate": "2018-01-15T10:00:00.0000000-07:00",
  "expirationDate": "2022-08-27T12:00:00.0000000-06:00",
  "credentialSubject": {
    "id": "...",
    "driversLicense": {
      "type": "Iso18013DriversLicense",
      "namespaces": {
        "org.iso.18013.5.1": {
          "term": "value"
        },
        "org.iso.18013.5.1.aamva": {
          "term": "value"
        }
      }
    }
  },
  "proof": {}
}
```

awoie commented:
> I've added namespaces in the JSON-LD context definition. The license now should have the following form:
> 
> ```json
> {
>   "@type": "Iso18013DriversLicense",
>   "namespaces": {
>     "org.iso.18013.5.1": {
>       "term": "value"
>     }
>   }
> }
> ```
> 
> This allows the definition of terms from other namespaces without risking ambiguities.
> 
> In a separate context, I've defined the AAMVA terms. Here is a VC example using both contexts:
> 
> ```json
> {
>   "@context": [
>     "https://www.w3.org/2018/credentials/v1",
>     "https://w3id.org/vdl/v1",
>     "https://w3id.org/vdl/aamva/v1"
>   ],
>   "issuer": "...",
>   "issuanceDate": "2018-01-15T10:00:00.0000000-07:00",
>   "expirationDate": "2022-08-27T12:00:00.0000000-06:00",
>   "credentialSubject": {
>     "id": "...",
>     "driversLicense": {
>       "type": "Iso18013DriversLicense",
>       "namespaces": {
>         "org.iso.18013.5.1": {
>           "term": "value"
>         },
>         "org.iso.18013.5.1.aamva": {
>           "term": "value"
>         }
>       }
>     }
>   },
>   "proof": {}
> }
> ```

can you provide an example of a VC with some concrete AAMVA and ISO properties such as DL number, height etc.?

timothee-haudebourg commented:
I removed the `namespaces` property to remove one level of nesting. Now a VC looks like this:
```json
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://w3id.org/vdl/v1",
    "https://w3id.org/vdl/aamva/v1"
  ],
  "issuer": "...",
  "issuanceDate": "2018-01-15T10:00:00.0000000-07:00",
  "expirationDate": "2022-08-27T12:00:00.0000000-06:00",
  "credentialSubject": {
    "id": "...",
    "driversLicense": {
      "type": "Iso18013DriversLicense",
      "org.iso.18013.5.1": {
        "family_name": "Haudebourg",
        "age_in_years": 29
      },
      "org.iso.18013.5.1.aamva": {
        "aka_family_name_v2": "Foo"
      }
    }
  },
  "proof": {}
}
```
I've also updated the example in `index.html` and the vocabulary table. I added a column specifying the expected datatype for each term. There is no table for the AAMVA terms, @awoie should I do that in the same section (5) or create a new section?

awoie commented:
> I removed the `namespaces` property to remove one level of nesting. Now a VC looks like this:
> 
> ```json
> {
>   "@context": [
>     "https://www.w3.org/2018/credentials/v1",
>     "https://w3id.org/vdl/v1",
>     "https://w3id.org/vdl/aamva/v1"
>   ],
>   "issuer": "...",
>   "issuanceDate": "2018-01-15T10:00:00.0000000-07:00",
>   "expirationDate": "2022-08-27T12:00:00.0000000-06:00",
>   "credentialSubject": {
>     "id": "...",
>     "driversLicense": {
>       "type": "Iso18013DriversLicense",
>       "org.iso.18013.5.1": {
>         "family_name": "Haudebourg",
>         "age_in_years": 29
>       },
>       "org.iso.18013.5.1.aamva": {
>         "aka_family_name_v2": "Foo"
>       }
>     }
>   },
>   "proof": {}
> }
> ```
> 
> I've also updated the example in `index.html` and the vocabulary table. I added a column specifying the expected datatype for each term. There is no table for the AAMVA terms, @awoie should I add one in the same section (5) or create a new section?

I would create a new table.

timothee-haudebourg commented:
@awoie I've added a section with a new table introducing the AAMVA vocabulary. I'm also using the `org.iso.18013.5.1.aamva` namespace in the example.

A few missing elements that should be nice to have in the end:
  - which terms are required/optional;
  - the RDF property (IRI) associated with each term (for instance, the `age_birth_year` term expands to the RDF property `https://w3id.org/vdl/org.iso.18013.5.1#ageBirthYear`);
  - an RDF definition of all the properties (an RDF Turtle file, for instance) that should be provided when dereferencing an RDF property IRI (e.g., `https://w3id.org/vdl/org.iso.18013.5.1#ageBirthYear`).

Should we do that here or in a future PR?

awoie commented:
> @awoie I've added a section with a new table introducing the AAMVA vocabulary. I'm also using the `org.iso.18013.5.1.aamva` namespace in the example.
> 
> A few missing elements that should be nice to have in the end:
> 
> * which terms are required/optional;
> * the RDF property (IRI) associated with each term (for instance, the `age_birth_year` term expands to the RDF property `https://w3id.org/vdl/org.iso.18013.5.1#ageBirthYear`);
> * an RDF definition of all the properties (an RDF Turtle file, for instance) that should be provided when dereferencing an RDF property IRI (e.g., `https://w3id.org/vdl/org.iso.18013.5.1#ageBirthYear`).
> 
> Should we do that here or in a future PR?

I think future is fine. It also depends a bit on the CCG and what their preference is.

awoie reviewed (approved):
lgtm. thank you very much! that is awesome.

timothee-haudebourg commented on index.html:
```diff
@@ -483,6 +494,133 @@ <h2>The Verifiable Driver's License Vocabulary</h2>
+      <p>
+This vocabulary extends the Verifiable Driver's License Core Vocabulary with
+terms for all data elements defined by the mDL Implementation Guidelines issued
+by the American Association of Motor Vehicle Administrators (AMVAA). It can be
```
Nice catch
